docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,240 @@
+
+We propose modifying this configuration struct to allow for a configuration
+file to be set for Conduit to read.
+
+```go
+// Config holds all configurable values for Conduit.
+type Config struct {
+	// File holds a Path to a configuration file that Conduit should load.
+	File struct {
+		Path string
+	}
+
+	DB struct {
+		Type   string
+		Badger struct {
+			Path string
+		}
+		Postgres struct {
+			ConnectionString string
+			Table            string
+		}
+	}
+
+	HTTP struct {
+		Address string
+	}
+	GRPC struct {
+		Address string
+	}
+
+	Log struct {
+		Level  string
+		Format string
+	}
+}
+```
+
+The above configuration is parsed at start time and provies a reference to our 
+pipeline configuration file. After parsing that file, we inject environment 
+variables into their corresponding templated locations, if any.
+
+```go
+// DB defines the interface for a key-value store.
+type DB interface {
+	// NewTransaction starts and returns a new transaction.
+	NewTransaction(ctx context.Context, update bool) (Transaction, context.Context, error)
+	// Close should flush any cached writes, release all resources and close the
+	// store. After Close is called other methods should return an error.
+	Close() error
+
+	Set(ctx context.Context, key string, value []byte) error
+	Get(ctx context.Context, key string) ([]byte, error)
+	GetKeys(ctx context.Context, prefix string) ([]string, error)
+}
+```
+
+### Config files as a database implementation
+
+`DB` is an interface for storing key-value pairs. Our `ConfigFileAsDatabase` 
+struct will store a reference to a parsed configuration file and map that to
+a key value representation for the `GetAll` call.
+
+A key part about `ConfigFileAsADatabase` is that it stubs out mutable
+operations, including transaction handling, enforcing our immutability
+during operation.
+
+We pass that `ConfigFileAsADatabase` to each new service at start time.
+Our ConfigFileAsDatabase will be bound with an in-memory database and passed
+to each new service when Conduit is starting.
+
+```go
+// NewService initializes and returns a pipeline Service.
+func NewService(logger log.CtxLogger, db database.DB) *Service {
+	return &Service{
+		logger:        logger.WithComponent("pipeline.Service"),
+		store:         NewStore(db),
+		instances:     make(map[string]*Instance),
+		instanceNames: make(map[string]bool),
+	}
+}
+
+svc := NewService(logger, &ConfigAsDatabase{})
+```
+
+### Hydration 
+
+Once the services are started, they will be empty. We must parse our config 
+file into a full pipeline by hydrating it with other necessary information 
+including environment information, connector configurations, and processors.
+
+This step will be responsible for populating the service's individual DB
+references with the resources necessary before `Init` is called after `Run`.
+
+### Immutability
+
+Conduit should be immutable at runtime. We could achieve this by stubbing 
+out all mutable methods as discussed in previous sections.
```

Comment:
How about pipeline status and the error (if any)? When Conduit starts, a pipeline's status is set to running (if it was stopped).

Reply:
Status would be considered part of the connector lifecycle and therefore mutable by necessity.